### DISCUSSION TITLE
What does Homebrew do that OpenSP can compile statically?

### DISCUSSION THREAD

**Discussion by wrobelda** [Homebrew/discussions#2207] [Writing Formulae/Casks]
For the context, I am working on a vcpkg port of OpenSP, which requires static compilation on macOS and Linux. See the discussion here: 
https://github.com/microsoft/vcpkg/discussions/20347

The problem I am facing is that I cannot compile it statically on Apple Silicon. A simple:

```
autoreconf --verbose --install --force
./configure --disable-debug --disable-dependency-tracking --enable-http --disable-doc-build --disable-shared --enable-static
make
```

gives:
```
/bin/sh ../libtool  --tag=CXX   --mode=link g++  -g -O2   -o onsgmls nsgmls.o SgmlsEventHandler.o RastEventHandler.o StringSet.o nsgmls_inst.o ../lib/libosp.la ../intl/libintl.a -liconv -lpthread
libtool: link: g++ -g -O2 -o onsgmls nsgmls.o SgmlsEventHandler.o RastEventHandler.o StringSet.o nsgmls_inst.o -Wl,-bind_at_load  ../lib/.libs/libosp.a ../intl/libintl.a -liconv -lpthread
ld: warning: ignoring file ../lib/.libs/libosp.a, building for _macOS-arm64_ but attempting to link with file built for _macOS-arm64_
Undefined symbols for architecture arm64:
```

Notice the puzzling error message: _building for **macOS-arm64** but attempting to link with file built for **macOS-arm64**_

However, if I `brew edit open-sp` and add `--disable-shared --enable-static` to the list of options, it will compile just fine. 

Can someone please point me to what does Brew do here that it manages to build it successfully? 

Note that I did see the brew's own `libtools` as dependency and I did try building manually with it being exported per the instructions (`export PATH="$(brew --prefix)/opt/libtool/libexec/gnubin:$PATH"`), yet it still fails. 

I am puzzled and would appreciate any help.

**Comment by Bo98** [comment]
Hard to tell - might need more of the build log. One suggestion I have is to make sure that GNU Binutils is not in your PATH (it has its own `ranlib` which conflicts with the macOS one).

Other than that, try take the full build log from your failing output and take the full build log of the Homebrew build (`~/Library/Logs/Homebrew/open-sp`) and compare the differences.

Homebrew also strips a few compiler flags (indicated by the `*.cc` logs), but it's unclear yet if that's anything to do with the differences you are seeing.

**Reply by wrobelda** [reply]
Thanks for your input, much appreciated!

Regarding binutils, I had checked that before.

Comparing the config files – yeah, definitely a good idea. I replicated the environment variables from Brew shell and got my build log as close to brew's as possible. The only thing that didn't work was prepending the PATH with `/opt/homebrew/Library/Homebrew/shims/mac/super`, which resulted in "C compiler cannot create executables" on my end. However, if I removed that from the Homebrew's env PATH, I was able to replicate the linker issue there, and the log files generated by it were 1:1 identical with my own. So, whatever makes cthe compilation possible is due to Brew's shim-ed compilers or other tools.

**Reply by wrobelda** [reply]
So it's the 'cmake' shim wrapper specifically that makes the compilation possible. I am now checking all the `HOMEBREW_` vars to find the actual reason.

**Reply by wrobelda** [reply]
OK, it's the `/opt/homebrew/lib` in `HOMEBREW_LIBRARY_PATHS` that is added to LDFLAGS by the `clang` wrapper, most likely caused by `liconv` or `lintl`
